cache_flat_mutation_reader.hh

```diff
@@ -730,20 +749,92 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
         }
     }
 
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
     if (_next_row_in_range) {
-        if (_next_row.range_tombstone_for_row() != _current_tombstone) [[unlikely]] {
-            auto tomb = _next_row.range_tombstone_for_row();
-            auto new_lower_bound = position_in_partition::before_key(_next_row.position());
-            clogger.trace("csm {}: rtc({}, {})", fmt::ptr(this), new_lower_bound, tomb);
-            push_mutation_fragment(mutation_fragment_v2(*_schema, _permit, range_tombstone_change(new_lower_bound, tomb)));
-            _lower_bound = std::move(new_lower_bound);
-            _current_tombstone = tomb;
-            _read_context.cache()._tracker.on_range_tombstone_read();
+        bool remove_row = false;
+
+        auto tomb_expired = [&](tombstone tomb) {
+            return (tomb && tomb.deletion_time < _gc_before);
+        };
+
+        if (_read_context.tombstone_gc_state() // do not compact rows when tombstone_gc_state is not set (used in some unit tests)
+            && _snp->at_latest_version()
+            && _snp->at_oldest_version()) {
+            if (!_next_row.dummy()) {
+                deletable_row& row = _next_row.latest_row();
+                tombstone range_tomb = _next_row.range_tombstone_for_row();
+                auto t = row.deleted_at();
+                t.apply(range_tomb);
+
+                auto row_tomb_expired = [&](row_tombstone tomb) {
+                    return (tomb && tomb.max_deletion_time() < _gc_before);
+                };
+
+                auto is_row_dead = [&](const deletable_row& row) {
+                    return (!row.marker().is_missing() && row.marker().is_dead(_read_time));
+                };
+
+                if (row_tomb_expired(t) || is_row_dead(row)) {
+                    can_gc_fn always_gc = [&](tombstone) { return true; };
+                    row_marker marker = row.marker();
+
+                    try {
+                        with_allocator(_snp->region().allocator(), [&] {
+                            row.compact_and_expire(*_schema, range_tomb, _read_time, always_gc, _gc_before, nullptr);
+                        });
+                        remove_row = row.empty();
+                    } catch (...) {
+                        row.marker() = marker;
```

Comment:
Why do we partially expire rows? Could we not expire rows atomically?

I like the solution of row_cache::_max_read_time, but don't see how you can maintain it easily. You'd need to keep reads sorted by _read_time and update any time the read that matches _max_read_time exits. It's doable but quite some work compared to keeping row processing atomic (which we already do everywhere).

Reply:
> Why do we partially expire rows? Could we not expire rows atomically?

It's about exceptions guarantees. I think expiration is in fact atomic today, (because there is nothing in `compact_and_expire` that can throw), but this isn't documented/guaranteed, so callers can assume weak guarantees only. We can just document that compact_and_expire has strong guarantees, and then the concern disappears. 

> I like the solution of row_cache::_max_read_time, but don't see how you can maintain it easily. You'd need to keep reads sorted by _read_time and update any time the read that matches _max_read_time exits. It's doable but quite some work compared to keeping row processing atomic (which we already do everywhere).

It sounds like you are thinking about min_read_time, but the proposal was max_read_time.

Reply:
compact_and_expire() may throw if it needs to re-serialize collections (`collection_mutation_description::serialize()`)

Reply:
> I like the solution of row_cache::_max_read_time, but don't see how you can maintain it easily. You'd need to keep reads sorted by _read_time and update any time the read that matches _max_read_time exits. It's doable but quite some work compared to keeping row processing atomic (which we already do everywhere).

It's enough to have something weaker. We don't need to provide monotonicity of read time across reads we execute, only snapshot isolation for rows (either expired or not expired). For that we only need to make sure that if a read touches a row which was partially expired, it will complete the expiration. For that, we need to keep the max expiry time point per row cache, and a flag to fire the compaction logic. This would be similar to sstable compaction, which doesn't care about read times of current reads, when it decides to expire stuff, all later reads will be affected. 

The problem with atomically expiring the row is that this logic runs deep inside an allocating section, and to retry we need to unwind the scope. When the reader retires, it may not execute the same logic, it may move on to reading from sstables for example, because the range is now discontinuous due to eviction. Maybe we could avoid that with some tricks, but it could also happen that before we get to the same row another exception is thrown earlier and fails the read without retry.

Reply:
Maybe we should avoid side effects inside allocating sections altogether, and maintain a log of things to do, and then those things, like compacting rows, can be easily executed atomically in their own allocating section. The disadvantage here is that efficiency will suffer.

Reply:
> The problem with atomically expiring the row is that this logic runs deep inside an allocating section, and to retry we need to unwind the scope.

How about this: copy the row, run expiration on the copy, swap the expired copy with the original. Won't that give us atomicity?

Reply:
I understand the problem but not completely get your idea about _max_read_time...

Reply:
I favor a two-pass approach so that any failures happen before changing anything.

Reply:
> Maybe we should avoid side effects inside (deep) allocating sections altogether, and maintain a log of things to do, and then those things, like compacting rows, can be easily executed atomically in their own allocating section. The disadvantage here is that efficiency will suffer.

Isn't the degradation only when we actually need to compact? Which is pretty rare.

Reply:
> > I like the solution of row_cache::_max_read_time, but don't see how you can maintain it easily. You'd need to keep reads sorted by _read_time and update any time the read that matches _max_read_time exits. It's doable but quite some work compared to keeping row processing atomic (which we already do everywhere).
> 
> It's enough to have something weaker. We don't need to provide monotonicity of read time across reads we execute, only snapshot isolation for rows (either expired or not expired). For that we only need to make sure that if a read touches a row which was partially expired, it will complete the expiration. For that, we need to keep the max expiry time point per row cache, and a flag to fire the compaction logic. This would be similar to sstable compaction, which doesn't care about read times of current reads, when it decides to expire stuff, all later reads will be affected.
> 
> The problem with atomically expiring the row is that this logic runs deep inside an allocating section, and to retry we need to unwind the scope. When the reader retires, it may not execute the same logic, it may move on to reading from sstables for example, because the range is now discontinuous due to eviction. Maybe we could avoid that with some tricks, but it could also happen that before we get to the same row another exception is thrown earlier and fails the read without retry.

Ok, thanks for the detailed explanation.

I already had a headache when starting to read this, but now it's worse. Anyway, I think the optimization should be for efficiency when nothing needs to be expired first, and simplicity second. Expiration is rare so it's okay if we waste some cycles there.

Reply:
Copying the row seems to be the simplest solution.

Reply:
do you mean solution like that?
```
with_allocator(_snp->region().allocator(), [&] {
    deletable_row row_copy(*_schema, row);
    row_copy.compact_and_expire(*_schema, t.tomb(), _read_time, always_gc, _gc_before, nullptr);
    std::swap(row, row_copy);
});
```

Reply:
> do you mean solution like that?

Yes.